components/script/docs/JS-Servos-only-GC.md

```diff
@@ -0,0 +1,396 @@
+most cases, lifetimes are [inferred][ti] and don't need to be written out in
+the source code. Inferred or not, the presence of lifetime information allows
+the compiler to reject use-after-free and other dangerous bugs.
+
+[lifetimes]: http://doc.rust-lang.org/guide-lifetimes.html XXX
+[ti]: http://en.wikipedia.org/wiki/Type_inference
+
+Not only do lifetimes protect Rust's built-in reference type, we can use them
+in our own data structures as well. `JSRef` is actually [defined][jsref-rs] as
+
+[jsref-rs]: XXX
+
+```rust
+pub struct JSRef<'a, T> {
+    ptr: NonZero<*const T>,
+    chain: PhantomData<&'a ()>,
+}
+```
+
+`T` is the familiar type variable, representing the type of DOM structure we're
+pointing to, e.g. `Window`. The somewhat odd syntax `'a` is a
+[lifetime variable][named-lifetime], representing the region of code in which
+that object is rooted. Crucially, this lets us write a [method][root-r] on
+`Root` with the following signature:
+
+[named-lifetime]: http://doc.rust-lang.org/guide-lifetimes.html#named-lifetimes
+[root-r]: XXX
+
+```rust
+pub fn r<'a>(&'a self) -> JSRef<'a, T> {
+    ...
+```
+
+What this syntax means is:
+
+- **`<'a>`**: 'for any lifetime `'a`',
+- **`(&'a self)`**: 'take a reference to a `Root` which is valid over lifetime `'a`',
+- **`-> JSRef<'a, T>`**: 'return a `JSRef` whose lifetime parameter is set to `'a`'.
+
+The final piece of the puzzle is that we put a [marker][phantom] in the `JSRef`
+type saying that it's only valid for the lifetime of a reference `&'a T`. This
+is how we extend the lifetime system to enforce our application-specific
+property about garbage collector rooting. If we try to compile something like
+this:
+
+[phantom]: http://doc.rust-lang.org/std/marker/struct.PhantomData.html
+
+```rust
+fn bogus_get_window<'a>(&self) -> JSRef<'a, Window> {
+    let window = self.window.root();
+    window.r()  // return the JSRef
+}
+```
+
+we get an error:
+
+XXX
+<pre class="sourceCode">`document.rs:199:9: 199:15 <span style="color: red">error:</span> <b>`window` does not live long enough</b>
+document.rs:199     window.root_ref()
+                    <span style="color: red">^~~~~~</span>
+document.rs:197:57: 200:6 <span style="color: green">note:</span> <b>reference must be valid for
+    the lifetime "a as defined on the block at 197:56...</b>
+document.rs:197 fn bogus_get_window<"a>(&amp;self) -> JSRef<"a, Window> {
+document.rs:198     let window = self.window.root();
+document.rs:199     window.root_ref()
+document.rs:200 }
+document.rs:197:57: 200:6 <span style="color: green">note:</span> <b>...but borrowed value is only
+    valid for the block at 197:56</b>
+document.rs:197 fn bogus_get_window<"a>(&amp;self) -> JSRef<"a, Window> {
+document.rs:198     let window = self.window.root();
+document.rs:199     window.root_ref()
+document.rs:200 }</pre>
+
+We also implement the [`Deref` trait][deref] for `JSRef<T>`. This allows us to
```

Comment:
Might want to mention that due to Rust's autoderef accessing the inner fields is usually not a problem,